### PR TITLE
Move ConfigOptions to core

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4349 

# Rationale for this change

Moves `ConfigOptions` to `common` so it can be used directly by the optimizer, planner, etc... greatly simplifying plumbing configuration around the place

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/Cargo.toml [hunk 1]
```diff
@@ -47,3 +47,4 @@ object_store = { version = "0.5.0", default-features = false, optional = true }
 parquet = { version = "29.0.0", default-features = false, optional = true }
 pyo3 = { version = "0.17.1", optional = true }
 sqlparser = "0.29"
+num_cpus = "1.13.0"
```
Comment (tustvold) [inline datafusion/common/Cargo.toml:50]: This is the only slight downside, but I'm inclined to think this isn't a big deal

File: datafusion/optimizer/src/optimizer.rs [hunk 2]
```diff
@@ -92,9 +92,12 @@ pub struct OptimizerContext {
 impl OptimizerContext {
     /// Create optimizer config
     pub fn new() -> Self {
+        let mut options = ConfigOptions::default();
+        options.optimizer.filter_null_join_keys = true;
```
Comment (tustvold) [inline datafusion/optimizer/src/optimizer.rs:96]: I double-checked this wasn't introduced by the recent config rework, and confirmed that it has always been the case that OptimizerContext defaults this to true, but ConfigOptions defaults it to false.

File: datafusion/common/src/config.rs [hunk 3]
```diff
@@ -17,7 +17,7 @@
 
 //! DataFusion Configuration Options
 
-use datafusion_common::{DataFusionError, Result};
+use crate::{DataFusionError, Result};
```
Comment (alamb) [inline datafusion/common/src/config.rs:20]: 🎉

File: datafusion/optimizer/src/filter_null_join_keys.rs [hunk 4]
```diff
@@ -43,8 +43,12 @@ impl OptimizerRule for FilterNullJoinKeys {
     fn try_optimize(
         &self,
         plan: &LogicalPlan,
-        _config: &dyn OptimizerConfig,
+        config: &dyn OptimizerConfig,
     ) -> Result<Option<LogicalPlan>> {
+        if !config.options().optimizer.filter_null_join_keys {
+            return Ok(None);
```
Comment (alamb) [inline datafusion/optimizer/src/filter_null_join_keys.rs:49]: maybe it is worth a `debug!` here if this is skipped

File: datafusion/optimizer/src/optimizer.rs [hunk 5]
```diff
@@ -91,28 +85,25 @@ pub struct OptimizerContext {
     /// Query execution start time that can be used to rewrite
     /// expressions such as `now()` to use a literal value instead
     query_execution_start_time: DateTime<Utc>,
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:87]: very nice

**Review by alamb — APPROVED** [review]
👨‍🍳 👌 

Looks very nice to me @tustvold  -- thank you

File: datafusion/optimizer/src/optimizer.rs (hunk 2 above)
Comment (andygrove) [reply datafusion/optimizer/src/optimizer.rs:96]: My intention was always to have this default to false so this looks like a bug

Comment (tustvold) [reply datafusion/optimizer/src/optimizer.rs:96]: `OptimizerContext` is now only used within tests, and so I am going to leave this for now. If you feel strongly I can file a follow up PR to change this, but it creates a non-trivial amount of test-churn

**Pull request merged**